keps/sig-architecture/4402-go-workspaces/README.md

```diff
@@ -0,0 +1,414 @@
+      `verify-foo.sh and `update-foo.sh` scripts) to use workspaces (often just
+      removing things like `GO111MODULE=off` or `-mod=mod` decorations).
+   1. Update gengo and kube-openapi versions in k/k (vendored).
+   1. Convert gengo-based tools (in staging/src/k8s.io/code-generator) to
+      gengo/v2 and kube-openapi/v2.
+   1. Fix any remaining tools which had dependencies on the gengo-based tools.
+   1. Remove detritus, including (but not limited to) run-in-gopath.sh,
+      references to GOPATH, workarounds for vendored packages, etc.
+
+Somewhere in there, we want to make deeper changes to gengo/v2 to simplify it.
+Some of the framework seemed clean at the time, but nearly a decade later, just
+makes the tools harder to understand.  This could happen after all of this
+work (repeating the sequence) or at the same time, making for larger but more
+complete reworking.
+
+### Risks and Mitigations
+
+There's not a clean way to do this.  Go workspaces DOES NOT work with our
+symlinks-in-vendor hack, and all of our tools do not work without it.  We have
+a LOT of tools and the hacks run DEEP.  We have a decade of legacy to scrub
+off.  Worse, many of our tools make dubious assumptions about the
+interchangeability of Go package paths (e.g. "example.com/foo/bar") and disk
+paths.  Those assumptions held as long as we used GOPATH, BARELY held with
+staging modules, and break totally in workspaces.
+
+This means that some of our ecosystem-facing APIs (codegen tools, in
+particular) *must* change.  We can mitigate some of the impact by creating "v2"
+of things like k8s.io/gengo (which powers many tools).  We might choose to do
+the same in kube-openapi, though its API is not changing, just the
+implementation.
+
+However, k8s.io/code-generator is more challenging.  It holds all of our tools
+and some scripts for invoking them, and is used by a non-trivial number of
+downstream projects (operators, custom API servers, etc).  Making a v2 of this
+(which itself lives in staging, and is subject to this KEP!!) is something we
+have not done before.  The simple answer is to rely on the fact that it was
+never really versioned before, and just make the changes.  We could move it all
+to a new "legacy-code-generator" staging repo and explain that it gets no
+further support, or we could just tag the last legacy release of
+k8s.io/code-generator and tell people to sync to that tag.  Same effect.
+
+Another risk is that this depends on
+[Go 1.22](https://tip.golang.org/doc/go1.22) and
+[vendoring in workspace mode](https://github.com/golang/go/issues/60056), which
+are not yet released. Go 1.22 is scheduled to release in February 2024, which
+seems like good timing for k8s 1.30, but has not yet been confirmed as
+plan-of-record.  As of Jan 13, 2024, people on the release team think it's
+likely that we will release k8s 1.30 on Go 1.22, but
+[that work](https://github.com/kubernetes/release/issues/3280) has not yet
+begun.
```

Comment:
Hidden in here in another possible risk I've started to think about:

If this merges close to code freeze, i has potential to disrupt the not-yet-merged PRs targeting the release. If we get lucky and 1.22 is available in early Feb and this merges well before the release, I don't see this being a problem. 

One possible way to think of merging this would be to treat proximity to code freeze as a taint.  We want to merge it, but since the release doesn't actually benefit from this change, it might be best to merge it well before 1.30, or just after the release when the branch first thaws...

Reply:
Roger, will note, though the collisions between this and other PRs should be minimal.